fix(animations): normalize final styles in buildStyles

dario-piotrowicz opened the pull request:
the final styles created in buildStyles lack normalization, meaning that pixel values remain as numbers (without "px") and so such properties fail to be correctly set/applied

Example: "width: 300" would be applied as "width": "300" (and thus ignored) instead of the correct "width": "300px"

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
At the end of animations numerical values are not normalized thus they are not saved correctly

Example with width, notice how the animation widths are ignored at the end:
![width-before](https://user-images.githubusercontent.com/61631103/124380354-2224e780-dcb4-11eb-9f6c-1bc75ee4811d.gif)

[stackblitz](https://stackblitz.com/edit/angular-animation-end-width-bug)


Issue Number: https://github.com/angular/angular/issues/24698


## What is the new behavior?

numerical values get normalized at animations end thus fixing the issue

Example after fix:
![width-after](https://user-images.githubusercontent.com/61631103/124380404-5bf5ee00-dcb4-11eb-9bd8-49bca05dc74a.gif)


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

dario-piotrowicz committed:
fix(animations): normalize final styles in buildStyles

the final styles created in buildStyles lack normalization, meaning that pixel values remain as numbers (without "px") and so such properties fail to be correctly set/applied

Example: "width: 300" is applies as "width": "300" (and thus ignored) instead of the correct "width": "300px"

JoostK commented on packages/animations/browser/src/util.ts:
```diff
@@ -159,14 +160,18 @@ function writeStyleAttribute(element: any) {
       }
-      element.style[camelProp] = styles[prop];
+      const rawValue = styles[prop];
+      const value = normalizer?.normalizeStyleValue(prop, prop, rawValue, []) ?? rawValue;
```
Since `normalizeStyleValue` does not return a nullable value, the `??` fallback is only here because of the usage of `.?` for `normalizer`. In that case a ternary expression would more closely indicate that, plus it generates less code as downleveling `??` and `.?` expands into multiple ternaries.

JoostK replied:
More importantly is the question of whether `normalizer` should really be optional. Why is the normalizer only needed when `setStyles` is called from the `onDestroy` callback, but not in ~10 other calls to `setStyles`? That sounds inconsistent to me.

dario-piotrowicz replied:
about the use of `??` and `.?` I already actually thought of that and wanted to put a comment
(this version seems quite clean to me but yeah, not as performant as others)

--- 

About the `normalizer` being options, I must admit that was a very naive think of me, you're totally right that it should not be, I will amend this! :slightly_smiling_face: 

(the thing is that the bug as only related to the end of animations so I focused on that, but yeah, most likely having the values not normalized in `setStyles` would effect other aspects of animations)

JoostK reviewed (commented):
Heya, thank you for taking a look at this. The animations package is not a simple one to get your hands dirty on.

It would be great if this could have a test. I think the best place for such a test would be in the integration suite located in `packages/core/test/animation`. I am also a bit concerned about the normalizer being optional.

dario-piotrowicz commented on packages/animations/browser/src/util.ts:
```diff
@@ -159,14 +160,18 @@ function writeStyleAttribute(element: any) {
       }
-      element.style[camelProp] = styles[prop];
+      const rawValue = styles[prop];
+      const value = normalizer?.normalizeStyleValue(prop, prop, rawValue, []) ?? rawValue;
```
having to call `normalizeStyleValue` with `prop`, `prop`, `rowValue` and `[]` is a bit annoying

but I noticed that there are already calls to the method with the first 2 arguments equal, so I guess that is not terrible

about passing `[]`, that is an array for string errors, I am just passing a new array and ignoring it as I think that there cannot be (new) errors, because if the animation is coming to an end the previous `setStyles` calls have already performed the checks

---

I thought of creating a new `AnimationStyleNormalizer` method which would only take the `prop` and `rawValue` (and would call `normalizeStyleValue` the same as I am doing here) but I figured I should not mess around to much with the existing classes and that this is preferable

dario-piotrowicz commented on packages/animations/browser/src/util.ts:
```diff
@@ -159,14 +160,18 @@ function writeStyleAttribute(element: any) {
 
-export function setStyles(element: any, styles: ɵStyleData, formerStyles?: {[key: string]: any}) {
+export function setStyles(
+    element: any, styles: ɵStyleData, formerStyles?: {[key: string]: any}|null,
```
I added the `|null` as I wanted to pass that in the `setStyles(rootElement, instruction.toStyles, null, this._normalizer)` call

I could have just done `setStyles(rootElement, instruction.toStyles, undefined, this._normalizer)`, but using `undefined` that way just looks wrong to me

dario-piotrowicz replied:
never mind this comment, I am going to make the `normalizer` not optional and thus there want be a need to pass `null` for the `formerStyles`

dario-piotrowicz commented:
> Heya, thank you for taking a look at this. The animations package is not a simple one to get your hands dirty on.
> 
> It would be great if this could have a test. I think the best place for such a test would be in the integration suite located in `packages/core/test/animation`. I am also a bit concerned about the normalizer being optional.

Thank you very much!!! :smiley: 

About the test, consider it done! :wink: 

Also the normalizer should definitely not be optional! you're totally right! I'm going to amend it :slightly_smiling_face:

dario-piotrowicz commented on packages/animations/browser/src/util.ts:
```diff
@@ -159,14 +160,16 @@ function writeStyleAttribute(element: any) {
         formerStyles[prop] = element.style[camelProp];
       }
-      element.style[camelProp] = styles[prop];
+      element.style[camelProp] = normalizer.normalizeStyleValue(prop, prop, styles[prop], []);
```
here I think I will need to pass a proper `errors` array and check the result

(as in: https://github.com/angular/angular/blob/c9b47f21822343286d056fd70aec882acc0ae788/packages/animations/browser/src/render/shared.ts#L82)

dario-piotrowicz commented on packages/core/test/animation/animations_with_web_animations_integration_spec.ts:
```diff
@@ -512,6 +512,48 @@ describe('animation integration tests using web animations', function() {
        expect(elm.style.getPropertyValue('position')).toEqual('fixed');
      });
+
+  it('should set normalized style property values on animation end', () => {
```
@JoostK hopefully this is sort of what you had in mind? :slightly_smiling_face:

JoostK replied:
I like it!

dario-piotrowicz replied:
thanks!!!! :smile:  :smile:  :smile:

dario-piotrowicz commented on packages/animations/browser/src/util.ts:
```diff
@@ -159,14 +160,18 @@ function writeStyleAttribute(element: any) {
         formerStyles[prop] = element.style[camelProp];
       }
-      element.style[camelProp] = styles[prop];
+      const normalizedProp = normalizer.normalizePropertyName(prop, []);
+      element.style[camelProp] =
+          normalizer.normalizeStyleValue(prop, normalizedProp, styles[prop], []);
```
@JoostK sorry for bothering you :cry: 

Could you advise on what to do with the errors here? 
(I am currently ignoring them since I just pass empty one-time arrays)

Should I gather the errors and throw as in: https://github.com/angular/angular/blob/c9b47f21822343286d056fd70aec882acc0ae788/packages/animations/browser/src/render/shared.ts#L82 ?

JoostK replied:
That is a very good question. There is currently one error reported for the case where a string without unit is provided for a dimensional property. I think reporting that is reasonable, however that would introduce a breaking change and limits this change from landing in patch.

Therefore, I think it's reasonable to ignore the errors for now such that the behavior is not affected. We can later revisit this choice in a major version if needed. It may depend a bit on how breaky it turns out in Google.

JoostK replied:
We may want to also have a test for the behavior we choose here; i.e. that a string without number bound to a dimension property applies the string verbatim without an error.

dario-piotrowicz replied:
About the breaking change thing, I understand and yeah I would really like it to just get this fixed now and add the error handling later :slightly_smiling_face: 

would you mind if I add a comment there? 
(something like: `// TODO: (try to) add error handling before new major release (introduces a breaking change)`)

dario-piotrowicz replied:
About the test, I actually tried to implement it so that but I actually think that the animations don't past a certain point if they don't conform to the standards

to be more clear, I tried to change my test and modified 300 to "300", when running it I got:
```
Error: Unable to animate due to the following errors:
         - Please provide a CSS unit value for width:300
```         
         
So I looked into it and it's the error thrown in `normalizeKeyframes` (the same I linked in my comment), that is called as part of the `TransitionAnimationEngine`'s `flush` method

So if I understood correctly, the `flush` happens still as part of the animation before the `destroy` to which I am mainly hooking into, so if there were actual errors in the values they would have been caught by `flush`, so I think that my ignoring those errors is actually correct as there may not be errors at this point, because if there were errors they would have already been caught and an error would have been thrown.....

at least that's my understanding of it (if you want I can keep looking into it and try to get more confidence in this)

does this checks out with you? what do you think?

JoostK replied:
I see, thanks for checking. Based on your comment, I realize that the styles do get normalized at some point, but apparently not for all usages. That is an inconsistency in how the styles are interpreted, which is what you are addressing by normalizing the values as they are written.

I was interested to know where the `instruction.toStyles` came from for which the normalization was necessary. These styles originate here:

https://github.com/angular/angular/blob/94e3d3ab70402629fb5882523bc2506d9551dac0/packages/animations/browser/src/dsl/animation_transition_factory.ts#L46-L49

Which calls into the following:

https://github.com/angular/angular/blob/94e3d3ab70402629fb5882523bc2506d9551dac0/packages/animations/browser/src/dsl/animation_transition_factory.ts#L104-L126

I feel like that would actually be the appropriate place to do the normalization as it's still during transaction building where errors can be reported, rather than during write (which creates the unfortunate situation where errors are no longer expected, hence the questions about how to deal with the errors). I think you can get the normalizer in `buildStyles` by adding it to `AnimationTransitionFactory.build`, `AnimationTransitionFactory.buildStyles` and `AnimationStateStyles.buildStyles`. It should then hopefully be possible to normalize the value that is written to `finalStyles`.

dario-piotrowicz replied:
Oh gosh, I am so sorry you had to look into things so much because of me :cry: 

I see what you mean, I will check it and try to place the normalization where it makes more sense :slightly_smiling_face: 

( again sorry for not having looked at this well enough :sob: )

JoostK replied:
Please, don't be sorry! This is how both of us get to learn something new :)

JoostK reviewed (commented):
I think this looks good now. I'm marking for presubmit to see how this behaves in Google.

dario-piotrowicz commented:
@JoostK of course your comment was spot on and I put the normalization in the `AnimationStateStyles`'s `buildStyles` method :slightly_smiling_face: 

Hopefully it should be all good now and I am not ignoring the error messages

I played around with it and all seems to be working fine, both correct and incorrect scenarios, camel cased styles and not etc...

dario-piotrowicz commented on packages/animations/browser/test/render/transition_animation_engine_spec.ts:
```diff
@@ -20,7 +20,7 @@ const DEFAULT_NAMESPACE_ID = 'id';
 const driver = new MockAnimationDriver();
 
-// these tests are only mean't to be run within the DOM
+// these tests are only meant to be run within the DOM
```
this is kind of unrelated, please let me know if I should put this into a separate commit or it's fine here :slightly_smiling_face:

dario-piotrowicz replied:
cool, I moved this to a separate commit then thanks! :slightly_smiling_face:

dario-piotrowicz commented:
@JoostK above my changes I noticed the `interpolateParams` call and I looked into it as I was curious about it and it's relationship with the normalization

I saw something interesting there, basically there the normalization breaks for numberical values, what I mean is the for example if I try to pass the number `20` as `font-size` I get an error: 
![Screenshot at 2021-07-10 15-27-18](https://user-images.githubusercontent.com/61631103/125166444-87d11200-e193-11eb-8247-8767f924ef2f.png)

I believe that must be related to the template params parsing, the param number must be converted to a string and that's why the normalization fails (the number `20` gets normalized to `"20px"`, but if you pass `"20"` that's just rightfully considered a wrong css value).

What's even more curious to me, is that even a numerical value provided in the default params (not in the template) also causes the same error (that could have been done on purpose to make the template and default params behaviours consistent).

( By the way, I did check and this is not caused by my changes, it already happens in the current angular version :stuck_out_tongue_winking_eye:  )

I think that that is unrelated here and that I should not touch the interpolated values in this PR but may be interesting to look into next and see if I can do anything about it (I guess it all depends on whether the template values can be correctly parsed as numbers)

Please let me know what you think about it :slightly_smiling_face:

dario-piotrowicz commented:
@JoostK? :cry:

JoostK commented:
Heya, sorry it took me a while to respond. I think I understand what you mean in your last comment about parameter normalization and I agree that it's a different issue, so no need to worry about that in this PR. I think I agree with your analysis that numeric parameters should not cause normalization failures due to a hidden conversion to strings.

dario-piotrowicz commented:
> Heya, sorry it took me a while to respond. I think I understand what you mean in your last comment about parameter normalization and I agree that it's a different issue, so no need to worry about that in this PR. I think I agree with your analysis that numeric parameters should not cause normalization failures due to a hidden conversion to strings.

Ah no worries, sorry for bothering you, I was just afraid you forgot about this PR :sweat_smile:
(I really really hope I didn't bother you :sweat: )

Thanks for the confirmation on the other thing, I will keep it in mind then :slightly_smiling_face:

dario-piotrowicz commented on packages/animations/browser/src/dsl/animation_transition_factory.ts:
```diff
@@ -118,6 +122,8 @@ export class AnimationStateStyles {
           if (val.length > 1) {
             val = interpolateParams(val, combinedParams, errors);
           }
+          const normalizedProp = normalizer.normalizePropertyName(prop, errors);
```
actually I did check that and I remember I did not put `normalizedProp` in the `finalStyles` as that changed the final output of the styles building, so I just wanted to keep my changes to a minimum and just fix what's needed without changing behaviour, I'll have another look and will put a more detailed comment so you can advise further :slightly_smiling_face:

dario-piotrowicz replied:
ok I think I got what you were saying :sweat_smile: , the create instruction will actually be different

(
for example this is an instruction we get without using `normalizedProp` here:
(note the `toStyles`'s font size)
![instructionWithout](https://user-images.githubusercontent.com/61631103/126035031-5a719519-a4c3-4a47-b58e-ae2497e9a55a.png)

and this is the instruction we get for the same animation if we do use the `normalizedProp` here:
![instructionWith](https://user-images.githubusercontent.com/61631103/126035058-42e06a31-e9b6-4ea0-8791-9da90648404f.png)
)

so we do actually get a different instruction, but then the `fromStyles` and `toStyles` are always used by either `setStyles` or `eraseStyles` which normalize the prop, so the difference in the instructions really doesn't make any difference

right? :slightly_smiling_face:

JoostK replied:
Indeed. There's a slight difference if both casings are present but I don't think it makes an observable difference in the end.

dario-piotrowicz replied:
ah ok thanks! I made the change :slightly_smiling_face:

dario-piotrowicz committed:
refactor(animations): fix typo in transition animations spec file

fix the typo "mean't" in the transition_animation_engine spec file
to "meant" (the typo is simply in a code comment so the change
does not have any impact)

dario-piotrowicz commented on packages/animations/browser/src/dsl/animation_trigger.ts:
```diff
@@ -11,14 +11,16 @@ import {copyStyles, interpolateParams} from '../util';
-export function buildTrigger(name: string, ast: TriggerAst): AnimationTrigger {
-  return new AnimationTrigger(name, ast);
+export function buildTrigger(
+    name: string, ast: TriggerAst, normalizer: AnimationStyleNormalizer): AnimationTrigger {
```
ah wow! sorry I did not notice that!

I find this strange thought... it looks like a pretty low level function to be in the public API :confused: 

What can we do about this?  :disappointed:

JoostK replied:
I would add a commit that removes the `@publicApi` marker from `buildTrigger` and `AnimationTrigger` at the beginning of this PR.

dario-piotrowicz replied:
Ok done thanks a lot! :slightly_smiling_face: 

___

In the comment message I've put a link to this comments thread so I just wanted to add some info here for anyone how could look into this:
- the `AnimationTrigger` is indeed not present in the [aio animations public api docs](https://angular.io/api/animations), there is an `AnimationTriggerMetadata` but it seems unrelated to the class
- in the codebase `AnimationTrigger` doesn't seem to be exported publicly (it does not surface to the animation's [`public_api.ts`](https://github.com/angular/angular/blob/master/packages/animations/public_api.ts))
- `buildTrigger` is also not present in the docs nor it seems to be exported publicly

dario-piotrowicz committed:
refactor(animations): remove publicApi annotations from triggers

the buildTrigger function and AnimationTrigger class are annotated
as publicApi which seems wrong and makes changes to the two
problematic, so they are being removed here

see: https://github.com/angular/angular/pull/42763/files#r671481902

dario-piotrowicz commented on packages/animations/browser/src/dsl/animation_trigger.ts:
```diff
@@ -7,29 +7,24 @@
-import {copyStyles, interpolateParams} from '../util';
-
-import {SequenceAst, StyleAst, TransitionAst, TriggerAst} from './animation_ast';
+import {SequenceAst, TransitionAst, TriggerAst} from './animation_ast';
```
I removed some unused imports here, I hope this if fine :slightly_smiling_face:

JoostK commented on packages/animations/browser/src/dsl/animation_trigger.ts:
```diff
@@ -15,17 +15,11 @@ import {AnimationStyleNormalizer} from './style_normalization/animation_style_no
 
 
 
-/**
```
I'd change the `fix(animations): remove publicApi annotations from triggers` commit to a `refactor(animations)`, as this change is not a user-facing fix (and does not need to go into the changelog)

dario-piotrowicz replied:
ah ok sorry :sweat: 

I'll change the other commit messages too then :slightly_smiling_face: 

thanks!!! :slightly_smiling_face:

JoostK commented:
I'm not an animations code owner so my approval is not sufficient. Hopefully @crisbeto can also have a look at this one.

dario-piotrowicz commented:
> I'm not an animations code owner so my approval is not sufficient. Hopefully @crisbeto can also have a look at this one.

Awesome!!! Thank you so very much for all your help!!!! :heart_eyes: :smiling_face_with_three_hearts:

AndrewKushnir commented:
FYI, I've started tests in [Google's codebase](https://test.corp.google.com/OCL:385334798:BASE:385334847:1626547608045:8f28cd22) (internal-only link) and will keep this thread updated once I get results.

crisbeto commented on packages/animations/browser/src/dsl/animation_transition_factory.ts:
```diff
@@ -101,7 +104,8 @@ function oneOrMoreTransitionsMatch(
   constructor(private styles: StyleAst, private defaultParams: {[key: string]: any}) {}
 
-  buildStyles(params: {[key: string]: any}, errors: string[]): ɵStyleData {
+  buildStyles(params: {[key: string]: any}, errors: string[], normalizer: AnimationStyleNormalizer):
```
Would it make sense to pass in the normalizer as a parameter to the `AnimationStateStyles` constructor? I don't imagine it's going to change between calls.

dario-piotrowicz replied:
yeah... that's a very good point....

sorry I missed that, I'll amend it now :sweat:

dario-piotrowicz replied:
all done thanks! :slightly_smiling_face:

crisbeto reviewed (approved):
Left one question, but otherwise LGTM.

JoostK commented:
@dario-piotrowicz You seem to have amended the last commit, which was the refactor to remove the imports. I think the goal was to create a fixup commit and squash that into the second commit.

dario-piotrowicz committed:
refactor(animations): remove unused animation trigger imports

in the animation_trigger.ts file there are unused imports
remove them to clean up the file

dario-piotrowicz commented:
> @dario-piotrowicz You seem to have amended the last commit, which was the refactor to remove the imports. I think the goal was to create a fixup commit and squash that into the second commit.

lol you noticed that immediately! :see_no_evil: 

force of habit to amend the last commit... sorry :sweat_smile: , should be all good now

AndrewKushnir commented:
FYI tests in Google's codebase (including a global run) went well.

// cc @JoostK

AndrewKushnir commented on packages/animations/browser/test/dsl/animation_trigger_spec.ts:
```diff
@@ -9,6 +9,7 @@
 import {animate, AnimationOptions, state, style, transition} from '@angular/animations';
 import {AnimationTransitionInstruction} from '@angular/animations/browser/src/dsl/animation_transition_instruction';
 import {AnimationTrigger} from '@angular/animations/browser/src/dsl/animation_trigger';
+import {NoopAnimationStyleNormalizer} from '../../src/dsl/style_normalization/animation_style_normalizer';
```
The `NoopAnimationStyleNormalizer` looks unused.

dario-piotrowicz replied:
sorry I forgot to remove it during the last refactoring :bow:

AndrewKushnir reviewed (approved):
@dario-piotrowicz thanks for the fix 👍
The changes look good, just left 1 minor comment on seemingly unused import.

AndrewKushnir commented:
@JoostK @crisbeto could you please have a final look and if everything is ok, we can probably add the PR to the merge queue?

JoostK commented:
Let's do it! I was waiting for CI to go green.
